### ISSUE TITLE
NullReferenceException in ValidateFile

### ISSUE THREAD

**Issue by Rockerby** [Shazwazza/Smidge#156]
We had an issue this morning where our site stopped working with a `NullReferenceException` in the `ValidateFile` function. It was running yesterday and there have been no code / file changes since then (also no server updates or restarts). The `ValidateFile` function is pretty small so assuming that the file that it's passing in along the way is null. The bundle in question is a CSS bundle.

My gut is telling me this is something to do with the application pool going idle as a restart fixes the issue. Relevant stack trace extract below:

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at Smidge.BundleFileSetGenerator.ValidateFile(IWebFile file)
   at Smidge.BundleFileSetGenerator.GetOrderedFileSet(IEnumerable`1 files, PreProcessPipeline pipeline)
   at Smidge.BundleFileSetGenerator.GetOrderedFileSet(Bundle bundle, PreProcessPipeline pipeline)
   at Smidge.SmidgeHelper.GenerateBundleUrlsAsync(String bundleName, String fileExt, Boolean debug)
   at Smidge.TagHelpers.SmidgeLinkTagHelper.ProcessAsync(TagHelperContext context, TagHelperOutput output)
   at Microsoft.AspNetCore.Razor.Runtime.TagHelpers.TagHelperRunner.<RunAsync>g__Awaited|0_0(Task task, TagHelperExecutionContext executionContext, Int32 i, Int32 count)
```  

For reference, this is an Umbraco 9.5.0 site (.NET 5), running Smidge 4.0.3 hosted on Windows Server 2019, IIS 10.

**Comment by nikro412956** [comment]
I have the same issue with the exact same conditions. It happened to 2 different websites so far. We might also need to report this to the Umbraco git repo so they can also look in to it.

**Comment by nikro412956** [comment]
@Rockerby What cache buster are you using? I am using the Timestamp cache buster. So i am curious which one you are using and if using the other cache buster can prevent this problem from occurring.

**Comment by Rockerby** [comment]
Hey @nikro412956 - this was using the `Version` cache buster type. I couldn't replicate this on a dev machine; purely on a live/staging environment.

**Comment by madsoulswe** [comment]
Yes, we got the same issue in production.

A couple of days ago the Igloo css-bundle returned 500 with the same exception. It was resoved after a restart.
Now the js-bundle got the same issue.

The site is using Version-cachebuster and I have disabled Smidge MemoryCache for now to se if it helps.

Umbraco 10.5.1 + Smdige 4.5
Windows server

**Comment by benehudson** [comment]
For what it's worth this issue is still present as of Umbraco 13.5.1

I've been having 11.2 and 12.2 sites randomly crashing for over a year now.  Restarting IIS fixes it every time.  I thought the issue was Umbraco, so I upgraded several of my sites to 13.5.1 last week.  I upgraded 5 sites before I could upgrade site number 6, the second site I upgraded to 13.5.1 crashed.  I recently attended Umbraco US Festival and from talking to other devs no one I talked to seemed to be experiencing these same random crashes. So when my 13.5.1 site crashed I dug into the logs and it points at a null reference inside Smidge.  I'm running on an Azure VM using Version cache buster as well.

**Comment by Shazwazza** [comment]
Is everyone reporting this issue occurring on your front-end or back office?

I'm trying to figure out how this could happen since it would mean that one of the file references for the bundle is null here https://github.com/Shazwazza/Smidge/blob/e022331d15c898c5876d4725004f746b3dfdcb05/src/Smidge.Core/BundleFileSetGenerator.cs#L113

The only way this could happen is if that file reference is null when creating the bundle initially.

Can anyone provide an example of how they are creating the bundle that is causing this issue? It could also occur if using view based declarations and RequireCss/RequireJs is used and passing in a null file.

I can add some null checks to these methods to try to avoid this.

**Comment by Shazwazza** [comment]
This is actually a duplicate of #205

**Comment by Shazwazza** [comment]
Have published version 4.5.0 https://github.com/Shazwazza/Smidge/releases/tag/v4.5.0 which eagerly adds null checks when creating a bundle with null files. Its not clear what the full underlying issue is but the file can only be null if it is registered as null when creating the bundle.

**Comment by benehudson** [comment]
@Shazwazza To answer your question I'm seeing this issue on the "front end". Smidge works great for me...most of the time.  What happens in my case is everything is working just fine and it's running on my server then occasionally (not every time) when the w3wp.exe recycles I'll get a null reference exception 500 error after the recycle (which effectively brings down my entire site as I'm using Smidge in my master template).  I've never seen it when developing locally, only have this rare but statically significant issue when my site is deployed on a server and immediately after a w3wp.exe recycle.

**Comment by Shazwazza** [comment]
@benehudson are you using view based declarations?

**Comment by benehudson** [comment]
@Shazwazza this is the code in my TemplateMaster.cshtml 

```
SmidgeHelper
        .CreateJsBundle("inline-js-bundle")
        .RequiresJs("~/scripts/slick.js", "~/scripts/jquery.fancybox.js", "~/scripts/isotope.images.loaded.js", "~/scripts/isotope.js", "~/scripts/jquery.google.reviews.js", "~/scripts/gsap.js", "~/scripts/gsap.scrolltrigger.js", "~/scripts/template.js");

    SmidgeHelper
        .CreateCssBundle("inline-css-bundle")
        .RequiresCss("~/css/slick.css", "~/css/slick-theme.css", "~/css/jquery.fancybox.css", "~/css/umbraco.block.grid.css", "~/css/template.css");
```
Then I use the taghelper to render the bundles.

**Comment by Shazwazza** [comment]
@benehudson where do you use the taghelper? Is that within your TemplateMaster.cshtml or somewhere else? View based declarations can be problematic, see https://github.com/Shazwazza/Smidge/issues/198#issuecomment-1646155044

**Comment by benehudson** [comment]
@Shazwazza yes the taghelpers are in my TemplateMaster.cshtml When you say "pre-defined bundles", you're saying moving the create bundle to the program.cs?

**Comment by Shazwazza** [comment]
Yep, like this https://github.com/Shazwazza/Smidge?tab=readme-ov-file#pre-defined-bundles

**Comment by Rockerby** [comment]
I haven't seen this issue for ages, but I can confirm I was getting this in view based declarations too. To be quite honest; I always thought that was the "right" way to do it.

Thanks for the pointer to the docs.

**Comment by fbpenalosa** [comment]
@Shazwazza  I know this issue has been a couple years old but I was wondering if there's been an update on this.

We've also been experiencing the same issues on our productions sites

`System.NullReferenceException: Object reference not set to an instance of an object.
   at Smidge.BundleFileSetGenerator.ValidateFile(IWebFile file)
   at Smidge.BundleFileSetGenerator.GetOrderedFileSet(IEnumerable`1 files, PreProcessPipeline pipeline)
   at Smidge.BundleFileSetGenerator.GetOrderedFileSet(Bundle bundle, PreProcessPipeline pipeline)
   at Smidge.Controllers.SmidgeController.Bundle(BundleRequestModel bundleModel)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(ActionContext actionContext, IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
--- End of stack trace from previous location ---`

No actual indicators that triggered it but the site would have to be restarted every time this occurs.

Our runtime minification settings is as follows:

`      "RuntimeMinification": {
        "UseInMemoryCache": true,
        "CacheBuster": "Version",
        "Version": ""
      },`

The Version property is blank here because we hooked it up to a startup process the pulls the value from a shared assembly. That way, the assets will be tied to version updates and clients wont have to be asked to refresh their end after a deploy

Below is part of the error that is happening when smidge is failing. I've also highlighted the version number generated that is tied to the runtime minification

<img width="1067" height="104" alt="Image" src="https://github.com/user-attachments/assets/74377feb-fc6f-43d6-84e0-3c97b78001d7" />

Were using:

Umbraco 12.3.6
USkinned 5.2.0

Thanks in advance

**Comment by fbpenalosa** [comment]
@Shazwazza  Its seems that an actual fix is being tied to this issue now. In relations to this, once the fix is ready, it is possible to obtain the fix by just updating the Smidge package without upgrading Umbraco?

<img width="907" height="297" alt="Image" src="https://github.com/user-attachments/assets/77f3d586-0b14-4f57-96d6-f97c0ec6ada3" />

**Comment by Shazwazza** [comment]
Closing this as reconciled.

As noted above, this was confirmed as a duplicate of #205. The runtime `NullReferenceException` in `BundleFileSetGenerator.ValidateFile` was fixed by merged PR #227, which skips null `IWebFile` entries with a warning instead of failing the whole bundle.

I also opened #232 to carry forward the only remaining useful part of draft #218: filtering null typed file entries during `BundleManager.Create(...)` so invalid bundle registrations do not persist nulls into bundle snapshots. #218 itself has been closed as superseded because its generator behavior conflicted with the already-merged #227 fix.
